eng/pipelines/templates/jobs/ci.yml

```diff
@@ -0,0 +1,113 @@
+parameters:
+  - name: ServiceDirectory
+    type: string
+    default: ""
+  - name: BuildTargetingString
+    type: string
+    default: "azure-*"
+  - name: BeforeTestSteps
+    type: object
+    default: []
+  - name: AfterTestSteps
+    type: object
+    default: []
+  - name: TestTimeoutInMinutes
+    type: number
+    default: 60
+  - name: BuildDocs
+    type: boolean
+    default: true
+  - name: TestProxy
+    type: boolean
+    default: false
+  - name: TestPipeline
+    type: boolean
+    default: false
+  - name: GenerateApiReviewForManualOnly
+    type: boolean
+    default: false
+  - name: BuildMatrix
+    type: object
+    default:
+      - pool:
+          os: linux
+          name: $(LINUXPOOL)
+          image: $(LINUXVMIMAGE)
+        RustToolchains: [stable, msrv, nightly]
+      - pool:
+          os: windows
+          name: $(WINDOWSPOOL)
+          image: $(WINDOWSVMIMAGE)
+        RustToolchains: [stable, msrv, nightly]
+      - pool:
+          os: macOS
+          name: $(MACPOOL)
+          vmImage: $(MACVMIMAGE)
+        RustToolchains: [stable, msrv, nightly]
+  - name: AnalyzeToolchains
+    type: object
+    default: [stable]
+
+jobs:
+  - ${{ each matrix in parameters.BuildMatrix }}:
+      - ${{ each RustToolchain in matrix.RustToolchains }}:
+          - job: Build_${{ matrix.pool.os }}_${{ RustToolchain }}
+            timeoutInMinutes: 90
+
+            pool: ${{ matrix.pool }}
+
+            steps:
+              - template: /eng/common/pipelines/templates/steps/sparse-checkout.yml
+                parameters:
+                  Paths: [/*]
+
+              - template: /eng/pipelines/templates/steps/test-packages.yml
+                parameters:
+                  ServiceDirectory: ${{ parameters.ServiceDirectory }}
+                  BuildTargetingString: ${{ parameters.BuildTargetingString }}
+                  RustToolchain: ${{ RustToolchain }}
+                  ArtifactSuffix: ${{ matrix.pool.os }}_${{ RustToolchain }}
+                  PublishArtifacts: ${{ and(eq(matrix.pool.os, 'linux'), eq(RustToolchain, 'stable')) }}
```

Comment:
This is the sort of conditioning that makes me nervous. Could instead add a boolean parameter like `publish: true` on that matrix slice and then check here: `${{ eq(matrix.publish, true) }}` (or maybe we don't need to use `eq` if empty strings/absent values are falsey already).

Reply:
I can expand it, but the matrix gets really verbose:

```yaml
  - name: BuildMatrix
    type: object
    default:
      - pool:
          os: linux
          name: $(LINUXPOOL)
          image: $(LINUXVMIMAGE)
        Toolchains:
          - name: stable
            publishArtifacts: true
          - name: msrv
          - name: nightly
      - pool:
          os: windows
          name: $(WINDOWSPOOL)
          image: $(WINDOWSVMIMAGE)
        Toolchains:
          - name: stable
          - name: msrv
          - name: nightly
      - pool:
          os: macOS
          name: $(MACPOOL)
          vmImage: $(MACVMIMAGE)
        Toolchains:
          - name: stable
          - name: msrv
          - name: nightly
```

Reply:
Would it be possible to set a variable then - something more obvious to change? GitHub Actions seems more capable in this instance, unfortunately.

Reply:
We can't make variable based as we use it in conditional yaml templating, so we're stuck with compile time parameters and expressions.

I just expanded the matrix and changed the parameter to:
```yaml
 PublishArtifacts: ${{ toolchain.publish }}
 ```